### ISSUE TITLE
vmware_deploy_ovf throws urlopen error The write operation timed out on large/multiple ovfs

### ISSUE THREAD

**Issue by pawelg007** [ansible/ansible#45276]
##### SUMMARY
While deploying large (around 10 GB) ovfs or four of them at once, it fails from time to time (20-30% fail rate), with "msg": "<urlopen error The write operation timed out>" on ESXI 6.0.0.

##### ISSUE TYPE
- Bug Report

##### COMPONENT NAME
vmware_deploy_ovf.py

##### ANSIBLE VERSION
```
ansible 2.6.1
  config file = /ansible/ansible.cfg
  configured module search path = ['/ansible/installation/modules', '/ansible/upgrade/modules', '/ansible/modules']
  ansible python module location = /usr/local/lib/python3.6/site-packages/ansible
  executable location = /usr/local/bin/ansible
  python version = 3.6.2 (default, Jun  5 2018, 19:03:23) [GCC 6.3.0 20170516]
```

##### CONFIGURATION
```
ANSIBLE_SSH_ARGS(/ansible/ansible.cfg) = -C -o ControlMaster=auto -o ControlPersist=10m
DEFAULT_ACTION_PLUGIN_PATH(/ansible/ansible.cfg) = ['/ansible/plugins/action']
DEFAULT_CONNECTION_PLUGIN_PATH(/ansible/ansible.cfg) = ['/ansible/plugins/connection']
DEFAULT_FORKS(/ansible/ansible.cfg) = 10
DEFAULT_MODULE_PATH(/ansible/ansible.cfg) = ['/ansible/installation/modules', '/ansible/upgrade/modules', '/ansible/modules']
DEFAULT_MODULE_UTILS_PATH(/ansible/ansible.cfg) = ['/ansible/module_utils']
DEFAULT_ROLES_PATH(/ansible/ansible.cfg) = ['/ansible/installation/roles', '/ansible/upgrade/roles', '/ansible/roles']
HOST_KEY_CHECKING(/ansible/ansible.cfg) = False
INVENTORY_UNPARSED_IS_FAILED(/ansible/ansible.cfg) = True
```

##### OS / ENVIRONMENT
Linux 4.9.13-201.fc25.x86_64 #1 SMP Tue Mar 7 23:47:11 UTC 2017 x86_64 GNU/Linux

##### STEPS TO REPRODUCE
Run deployment or large/multiple ovf files in a loop.

```yaml
- name: deploy ovf
  delegate_to: "{{ file_broker_host }}"
  vmware_deploy_ovf:
    hostname: "{{ esxi_host }}"
    username: "{{ esxi_user }}"
    password: "{{ esxi_password }}"
    validate_certs: "{{ ssl_verify }}"

    name: "{{ hostname }}"
    ovf: "{{ ovf_path }}/{{ ovf }}"
    allow_duplicates: False
    power_on: False
    wait: False
    resource_pool: "{{ ovf_resource_pool }}"
    datastore: "{{ esxi_datastore }}"
```

##### EXPECTED RESULTS
success

##### ACTUAL RESULTS
```paste below
2018-09-06 01:23:25,298 p=2346 u=root |  TASK [deploy : deploy ovf] *******************************************************************************************************************************************************************************************************************
2018-09-06 01:23:25,299 p=2346 u=root |  task path: /ansible/roles/deploy/tasks/main.yml:16
2018-09-06 01:23:25,544 p=2346 u=root |  Using module file /ansible/modules/vmware/vmware_deploy_ovf.py
2018-09-06 01:23:25,545 p=2346 u=root |  Using module file /ansible/modules/vmware/vmware_deploy_ovf.py
2018-09-06 01:23:25,556 p=2346 u=root |  Using module file /ansible/modules/vmware/vmware_deploy_ovf.py
2018-09-06 01:23:25,548 p=2346 u=root |  Using module file /ansible/modules/vmware/vmware_deploy_ovf.py
2018-09-06 01:33:57,053 p=2346 u=root |  The full traceback is:
  File "/tmp/ansible_w5vqdbph/ansible_module_vmware_deploy_ovf.py", line 252, in run
    self._open_url()
  File "/tmp/ansible_w5vqdbph/ansible_module_vmware_deploy_ovf.py", line 240, in _open_url
    open_url(self.url, data=self.f, validate_certs=self.validate_certs, **self._request_opts())
  File "/tmp/ansible_w5vqdbph/ansible_modlib.zip/ansible/module_utils/urls.py", line 958, in open_url
    r = urllib_request.urlopen(*urlopen_args)
  File "/usr/local/lib/python3.6/urllib/request.py", line 223, in urlopen
    return opener.open(url, data, timeout)
  File "/usr/local/lib/python3.6/urllib/request.py", line 526, in open
    response = self._open(req, data)
  File "/usr/local/lib/python3.6/urllib/request.py", line 544, in _open
    '_open', req)
  File "/usr/local/lib/python3.6/urllib/request.py", line 504, in _call_chain
    result = func(*args)
  File "/tmp/ansible_w5vqdbph/ansible_modlib.zip/ansible/module_utils/urls.py", line 358, in https_open
    return self.do_open(self._build_https_connection, req)
  File "/usr/local/lib/python3.6/urllib/request.py", line 1320, in do_open
    raise URLError(err)

2018-09-06 01:33:57,053 p=2346 u=root |  fatal: [ae2 -> localhost]: FAILED! => {
    "changed": false,
    "invocation": {
        "module_args": {
            "allow_duplicates": false,
            "datacenter": "ha-datacenter",
            "datastore": "datastore1",
            "deployment_option": null,
            "disk_provisioning": "thin",
            "fail_on_spec_warnings": false,
            "folder": null,
            "hostname": "172.28.180.23",
            "name": "AE2-1A-1",
            "networks": {
                "VM Network": "VM Network"
            },
            "ovf": "/ansible/ovf/Win2012R2-OVF-01.03.07.01-02-OsReconfiguratorOff.ovf",
            "password": "VALUE_SPECIFIED_IN_NO_LOG_PARAMETER",
            "port": 443,
            "power_on": false,
            "properties": null,
            "resource_pool": "Resources",
            "username": "root",
            "validate_certs": false,
            "wait": false,
            "wait_for_ip_address": false
        }
    },
    "msg": "<urlopen error The write operation timed out>"
}
```

**Comment by ansibot** [comment]
Files identified in the description:
* [lib/ansible/modules/cloud/vmware/vmware_deploy_ovf.py](https://github.com/ansible/ansible/blob/devel/lib/ansible/modules/cloud/vmware/vmware_deploy_ovf.py)

If these files are inaccurate, please update the `component name` section of the description or use the `!component` bot command.

[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by ansibot** [comment]
cc @Akasurde @dav1x @sivel @warthog9
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by ansibot** [comment]
Hi @pawelg007,

Thank you for the issue, just so you are aware we have a dedicated Working Group for vmware.
You can find other people interested in this in #ansible-vmware on Freenode IRC
For more information about communities, meetings and agendas see https://github.com/ansible/community

[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by albers** [comment]
I've also seen this when deploying 20 OVA files with ~300 GB each in parallel on vCenter/ESXi 6.7.0

**Comment by sivel** [comment]
I'm not positive that there is really a solution to this.  This is going to depend on the network stability and available bandwidth.  Many large simultaneous transfers like this are likely to cause issues.

Limiting the concurrency is likely best here.  Such as using `serial` to limit to a 1 or 2 at a time, based on reliability and bandwidth.

**Comment by ansibot** [comment]
cc @ckotte
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by ansibot** [comment]
cc @lparkes
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by ansibot** [comment]
cc @Tomorrow9
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by ansibot** [comment]
cc @goneri
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by ansibot** [comment]
cc @pgbidkar
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by dapiced** [comment]
I have exactly de same issue when doing "vmware_deploy_ovf " in a loop

`- name: Importer le template vmware sur les autres vcenter app et ges
  vmware_deploy_ovf:
    validate_certs: False
    username:   "{{ vcenter_ad_rw.username }}"
    password:   "{{ vcenter_ad_rw.password }}"
    datacenter: "{{ item.dc }}"
    datastore:  "{{ item.ds }}"
    cluster:    "{{ item.cl }}"
    networks:   
      vxw-dvs-66-universalwire-3-sid-20002-ulsgestionb2drv01: "{{ item.nw }}"   
    hostname:   "{{ item.vc }}"    
    folder:     "{{ item.dc }}/vm/{{ fe_win_maj_template_vw__folder_drv }}"
    name:       "{{ fe_win_maj_template_vw__ProchainTemplateCourant }}"
    ovf:        "/var/tmp/win2016/{{ fe_win_maj_template_vw__ProchainTemplateCourant }}.ova"  
    power_on: no  
  vars:
    ansible_python_interpreter: /var/endpoint27/bin/python  
  environment:
    LD_LIBRARY_PATH: /opt/rh/python27/root/usr/lib64    
  tags: import_template_to
  loop: "{{ fe_win_maj_template_vw__vcdc_lst }}"`

Any idea to solve the issue wihout loop ? something to add to the task like this ?

async: 1000
poll: 0

OR 

wait_for: timeout=5000

thanks
Dominic

**Comment by Belabacsi** [comment]
Hello Dominic!

Same issue here, using vmware_deploy_ovf in loop. Did You find a solution for this issue?
Many thanks!
Best Regards,
Belabacsi

**Comment by dapiced** [comment]
> Hello Dominic!
> 
> Same issue here, using vmware_deploy_ovf in loop. Did You find a solution for this issue?
> Many thanks!
> Best Regards,
> Belabacsi

The issue seem to be in the ansible loop itself, so i simply repeat the task the number of time i need (without the loop function).

 Other solution can be to do a yml file loop instead a task loop...

**Comment by Belabacsi** [comment]
Hello Dominic!

Thank You for the quick feedback. Unfortunately it seems my situation is a little but different, because the symptom is the same also in case of running task without loop. I'm trying to import OVA files with 450 and 580 MB in size, so they are far from gigabytes and the symptom is the same... The first one works, the second one fails.
I have also checked if there are enough resources available on destination ESXi for hosting the VM (deploying and importing the VM manually was successful). What I'm thinking now about is one OVA contains a thick HDD instead of thin...

I have found a blog entry describing similar issue with another VMware Ansible module and it seems the below workaround works for them:
https://blog.storagemadeeasy.com/how-to-use-ansible-to-automate-vmware-ovas/
I'll try this method and let You know.

**Comment by dapiced** [comment]
> Hello Dominic!
> 
> Thank You for the quick feedback. Unfortunately it seems my situation is a little bit different, because the symptom is the same also in case of running task without loop. I'm trying to import OVA files with 450 and 580 MB in size, so they are far from gigabytes and the symptom is the same... The first one works, the second one fails.
> It is also important to note that, my error message says "read timeout" instead of write.
> I have also checked if there are enough resources available on destination ESXi for hosting the VM (deploying and importing the second larger VM manually were successful). What I'm thinking now about is one OVA contains a thick HDD instead of thin...
> 
> I have found a blog entry describing similar issue with another VMware Ansible module and it seems the below workaround works for them:
> https://blog.storagemadeeasy.com/how-to-use-ansible-to-automate-vmware-ovas/
> I'll try this method and let You know.

Hi,  interesting blog, in my case sometimes it fail when i import a OVA to a Vcenter with the  "vmware_deploy_ovf" module  and i got a write timeout error.  Indeed, the "vmware_deploy_ovf" module seem unstable. I don't think it is on the vmware side the issue... Maybe we will need to check the ansible tower log  to see what going on..... just hope the next ansible release will fix this issue.  let us know if the ovftoof solve your issue !  thanks ! Dominic

**Comment by Belabacsi** [comment]
Hello Dominic!

The below workaround works perfectly :-) Please note that, while the vmware_deploy_ovf module is idempotent, the ovftool method is not. This means, You can only deploy the VM(s) if they do not exist in vCenter. Running the playbook again, vCenter will drop an error complaining about that, the VM(s) already exist. We need to define another task to check if vCenter already has the VM(s) defined. If not, run ovftool, if yes skip deploying and power on the modules with vmware_guest_powerstate module.

```
        - name: "Deploy NS-ADM VM from OVF file"
          shell: | 
            /usr/bin/ovftool \
            --acceptAllEulas \
            --name="{{ item.vm_name }}" \
            --datastore="{{ vmware_vcenter_datastore }}" \
            --net:"VM Network"="{{ vmware_vcenter_default_network }}" \
            --X:logFile=/tmp/ovftool-log.txt --X:logLevel=verbose \
            --diskMode=thin \
            "{{ nsadm_images_path }}/{{ nsadm_image_extract_to }}/{{ nsadm_image_extract_to }}.ovf" \
            'vi://{{ vmware_vcenter_username }}:{{ vmware_vcenter_password }}@{{ vmware_vcenter_hostname }}/{{ vmware_vcenter_datacenter }}/host/{{ vmware_vcenter_cluster }}/'
          register: vmware_vm_deploy_ovf_result
          with_items: "{{ vmware_nsadm_vm_2_deploy }}"
```
I have followed [this quide](https://www.virtual-odyssey.com/2017/11/26/install-vmware-ovftool-ubuntu/) to install ovftool, You can download ovftool [here](https://code.vmware.com/web/tool/4.3.0/ovf).
Best Regards,
Belabacsi

**Comment by ansibot** [comment]
Thank you very much for your interest in Ansible. Ansible has migrated much of the content into separate repositories to allow for more rapid, independent development. We are closing this issue/PR because this content has been moved to one or more collection repositories.

* lib/ansible/modules/cloud/vmware/vmware_deploy_ovf.py -> https://galaxy.ansible.com/community/vmware

For further information, please see:
https://github.com/ansible/ansibullbot/blob/master/docs/collection_migration.md
